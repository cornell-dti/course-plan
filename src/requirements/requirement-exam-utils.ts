import examData, { ExamFulfillment, ExamFulfillments } from './data/exams/ExamCredit';
import { NO_FULFILLMENTS_COURSE_ID } from './data/constants';

type ExamType = keyof typeof examData;
type ExamTaken = {
  readonly examType: ExamType;
  readonly subject: string;
  readonly score: number;
};
export type ExamsTaken = ExamTaken[];
type ExamSubjects = Record<ExamType, string[]>;

const getExamFulfillment = (userExam: ExamTaken): ExamFulfillment | undefined => {
  const exam = examData[userExam.examType][userExam.subject];
  const fulfillment = exam.reduce((prev: ExamFulfillment | undefined, curr: ExamFulfillment) => {
    // check if exam name matches and score is high enough
    if (userExam.score >= curr.minimumScore) {
      // update exam variable if this exam has a higher minimum score
      if (!prev || prev.minimumScore < curr.minimumScore) {
        return curr;
      }
    }
    return prev;
  }, undefined);
  return fulfillment;
};

export const examsTakenToExamCourses = (exams: ExamsTaken): CourseTaken[] => {
  const examCourses: CourseTaken[] = [];
  exams.forEach(exam => {
    // match exam to fulfillment
    const fulfillment = getExamFulfillment(exam);

    // generate exam course
    const examName = `${exam.examType} ${exam.subject}`;
    if (fulfillment) {
      const { courseId, credits } = fulfillment;
      examCourses.push({
        courseId,
        uniqueId: examName,
        code: examName,
        credits,
      });
    } else {
      examCourses.push({
        courseId: NO_FULFILLMENTS_COURSE_ID,
        uniqueId: examName,
        code: examName,
        credits: 0,
      });
    }
  });
  return examCourses;
};

export default function userDataToExamCourses(user: AppOnboardingData): CourseTaken[] {
  const examsTaken = user.exam.map(({ type: examType, subject, score }) => ({
    examType,
    subject,
    score,
  }));
  return examsTakenToExamCourses(examsTaken);
}

const toSubjects = (data: ExamFulfillments) => {
  const subjects = [...new Set(Object.keys(data))];
  subjects.sort();
  return subjects;
};

export const examSubjects: ExamSubjects = {
  AP: toSubjects(examData.AP),
  IB: toSubjects(examData.IB),
};

export const getExamCredit = (examTaken: FirestoreAPIBExam): number => {
  const exam = examData[examTaken.type][examTaken.subject];
  const mostPossibleCredit = exam.reduce((credit, fulfillment) => {
    if (examTaken.score >= fulfillment.minimumScore) {
      return Math.max(credit, fulfillment.credits);
    }
    return credit;
  }, 0);
  return mostPossibleCredit;
};
