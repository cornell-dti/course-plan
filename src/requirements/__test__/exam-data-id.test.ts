import { examSubjects } from '../requirement-exam-utils';

const examIds = [
  ...examSubjects.AP.map(subject => `AP ${subject}`),
  ...examSubjects.IB.map(subject => `IB ${subject}`),
  // ...examSubjects.CASE.map(subject => `CASE ${subject}`),
].sort((a, b) => a.localeCompare(b));

it('ID of the exams are unchanged.', () => {
  /**
   * What should you do when this test fails:
   * a. Unbreak the test. To update the snapshot, run `npm run test -- -u`.
   * b. Write a migration script for any exam unique ids that changed.
   */

  expect(examIds).toMatchSnapshot();
});

it('No duplicate exam unique ID.', () => {
  const set = new Set<string>();
  examIds.forEach(id => {
    if (set.has(id)) {
      fail(`Detected duplicated exam unique ID: ${id}`);
    } else {
      set.add(id);
    }
  });
});
