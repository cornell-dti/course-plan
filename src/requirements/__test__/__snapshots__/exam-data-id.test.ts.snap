// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ID of the exams are unchanged. 1`] = `
Array [
  "AP Biology",
  "AP Chemistry",
  "AP Computer Science A",
  "AP English Language and Composition",
  "AP English Literature and Composition",
  "AP French Language",
  "AP French Literature",
  "AP Italian Language",
  "AP Italian Literature",
  "AP Macroeconomics",
  "AP Mathematics AB",
  "AP Mathematics BC (Engineering)",
  "AP Mathematics BC (Non-Engineering)",
  "AP Microeconomics",
  "AP Physics C-Electricity & Magnetism",
  "AP Physics C-Mechanics",
  "AP Physics I",
  "AP Physics II",
  "AP Psychology",
  "AP Spanish Language",
  "AP Spanish Literature",
  "AP Statistics",
  "IB Chemical and Physical Systems",
  "IB Chemistry",
  "IB Computer Science",
  "IB Economics",
  "IB English Language and Literature",
  "IB English Literature A",
  "IB Mathematics",
  "IB Physical Science",
  "IB Physics",
]
`;
